File: ben/prolog/P31-P41-arithmetic/p34.pl
% Calculate Euler's totient function phi(m).

:- ensure_loaded(p33).
:- arithmetic_function(totient_phi/1).

totient_phi(1, 1) :- !.
totient_phi(X, N) :-
    X1 is X - 1, 
    numlist(1, X1, L),
    partition(coprime(X), L, S, _), 
    length(S, N).  

File: ben/haskell/Q31-41-arithmetic/p34.hs
{-
Calculate Euler's totient function phi(m).

Euler's so-called totient function phi(m) is defined as the number of positive integers r (1 <= r < m) that are coprime to m.

Example: m = 10: r = 1,3,7,9; thus phi(m) = 4. Note the special case: phi(1) = 1.
-}

coprime :: Int -> Int -> Bool
coprime a b = gcd a b == 1

totient :: Int -> Int
totient a = length [b | b <- [1..a-1], 
                        coprime a b]
